libmamba/tests/src/util/test_iterator.cpp

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) 2023, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#include <algorithm>
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <doctest/doctest.h>
+
+#include "mamba/util/iterator.hpp"
+
+using namespace mamba::util;
+
+namespace
+{
+    bool greater_than_10(int i)
+    {
+        return i > 10;
+    }
+
+    bool second_greater_than_10(const std::pair<std::string, int>& p)
+    {
+        return p.second > 10;
+    }
+
+    auto get_test_data()
+    {
+        std::vector<int> input_sequence = { 1, 12, 2, 3, 14, 4, 18, 20, 4 };
+        std::vector<int> result_sequence = { 12, 14, 18, 20 };
+        std::map<std::string, int> input_map =
```

Comment:
Except if you need the values to be stable in memory, prefer unordered_map by default, or flat_map when there is such low number of small (in memory) values. Here it will help with test performance.

Reply:
`flat_map` is C++23 only, and `unordered_map` provide forward iterators, not bidirectional ones. That said, I should probably improve the `filter_iterator` so that it can handle forward iterators too.